src/StackExchange.Redis/PhysicalConnection.cs

```diff
@@ -1432,6 +1432,17 @@ internal async ValueTask<bool> ConnectedAsync(Socket socket, LogProxy? log, Sock
                     {
                         try
                         {
+#if NETCOREAPP3_1_OR_GREATER
+                            var configOptions = config.SslClientAuthenticationOptions?.Invoke(host);
+                            if (configOptions is not null)
+                            {
+                                await ssl.AuthenticateAsClientAsync(configOptions);
+                            }
+                            else
+                            {
+                                ssl.AuthenticateAsClient(host, config.SslProtocols, config.CheckCertificateRevocation);
+                            }
+#endif
                             ssl.AuthenticateAsClient(host, config.SslProtocols, config.CheckCertificateRevocation);
```

Comment:
@NickCraver 
This pre-processor directive needs a **else** block, otherwise runs ssl.AuthenticateAsClient **twice** in NETCOREAPP3_1_OR_GREATER

Something like this:

``` csharp
#if NETCOREAPP3_1_OR_GREATER
                            var configOptions = config.SslClientAuthenticationOptions?.Invoke(host);
                            if (configOptions is not null)
                            {
                                await ssl.AuthenticateAsClientAsync(configOptions);
                            }
                            else
                            {
                                ssl.AuthenticateAsClient(host, config.SslProtocols, config.CheckCertificateRevocation);
                            }
#else 
                   ssl.AuthenticateAsClient(host, config.SslProtocols, config.CheckCertificateRevocation);
#endif
```

Reply:
See comment above (and latest code) :)

Reply:
@NickCraver 

So far things are looking good in my environment. I was able to connect to a Redis instance with restricted ciphers from a .NET 6.0 client, both from Windows and Linux.

I needed to wrap the call to the CipherSuitesPolicy constructor in a try/catch because that will fail on Windows. 

Please let me know if the below snippet is how you indent this new API to be used.

``` csharp
var redisConfig = ConfigurationOptions.Parse(_configuration.GetConnectionString("Redis"));

redisConfig.SslClientAuthenticationOptions = (host) =>
{
	var sslOptions = new SslClientAuthenticationOptions
	{
		TargetHost = host
	};

	try
	{
		sslOptions.CipherSuitesPolicy = new CipherSuitesPolicy(new[]
		{
			TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
			// add more allowed ciphers here ...
		});
	}
	catch (PlatformNotSupportedException)
	{
		// Ignore: The CipherSuitesPolicy constructor is expected to fail on Windows. It only works on Linux with OpenSSL 1.1.1 or higher or MacOS
		// https://docs.microsoft.com/en-us/dotnet/api/system.net.security.ciphersuitespolicy?view=net-6.0
		// if (Environment.OSVersion.Platform == PlatformID.Unix) could also work here
	}

	return sslOptions;
};

var redis = ConnectionMultiplexer.Connect(redisConfig);
```

Reply:
Maybe! Alternatively, you could only set the property at all on Linux - thoughts?

Reply:
As this is providing a lot of flexibility, I don't think there's any need for extra customization here other than wrapping up and re-throwing any errors that emanate from the SslClientOptions invocation. There are probably other permutations where that invocation could throw an exception, they should be pretty transparent though, and this is flagged as a somewhat advanced use-case. All the ConfigurationOptions really needs to worry about is if the ssl-config delegate responded with valid options or null, it's really none of its business how those were derived as those are the responsibility of the consumer.

Reply:
Agreed. My use case only needs to set the SslClientAuthenticationOptions when running on Linux.

So I'm going with something like below.


``` csharp
var redisConfig = ConfigurationOptions.Parse(_configuration.GetConnectionString("Redis"));

if (Environment.OSVersion.Platform == PlatformID.Unix)
{
	redisConfig.SslClientAuthenticationOptions = (host) =>
		new SslClientAuthenticationOptions
		{
			TargetHost = host,
			CipherSuitesPolicy = new CipherSuitesPolicy(new[]
				{
					TlsCipherSuite.TLS_ECDHE_RSA_WITH_AES_128_CBC_SHA256,
					// add more allowed ciphers here ...
				})
		};
}

var redis = ConnectionMultiplexer.Connect(redisConfig);
```

Reply:
Oh yeah @slorello89 makes a good point, returning null is equivalent so however you want to structure the code there works. As long as this is working, it has legs and I very much appreciate the testing! :)